spec/unit/pops/types/p_uri_type_spec.rb

```diff
@@ -0,0 +1,191 @@
+        CODE
+        expect(eval_and_collect_notices(code)).to eq([
+          "URI('urn:a:b:c')",
+          "URI[{'scheme' => 'urn', 'opaque' => 'a:b:c'}]"
+        ])
+      end
+
+      it 'can be created from a hash' do
+        code = <<-CODE
+            $o = URI(scheme => 'https', host => 'example.com', path => '/a/b')
+            notice(String($o, '%p'))
+            notice(type($o))
+        CODE
+        expect(eval_and_collect_notices(code)).to eq([
+          "URI('https://example.com/a/b')",
+          "URI[{'scheme' => 'https', 'host' => 'example.com', 'path' => '/a/b'}]"
+        ])
+      end
+
+      it 'which is opaque, can be created from a hash' do
+        code = <<-CODE
+            $o = URI(scheme => 'urn', opaque => 'a:b:c')
+            notice(String($o, '%p'))
+            notice(type($o))
+        CODE
+        expect(eval_and_collect_notices(code)).to eq([
+          "URI('urn:a:b:c')",
+          "URI[{'scheme' => 'urn', 'opaque' => 'a:b:c'}]"
+        ])
+      end
+
+      it 'is an instance of its type' do
+        code = <<-CODE
+            $o = URI('https://example.com/a/b')
+            notice($o =~ type($o))
+        CODE
+        expect(eval_and_collect_notices(code)).to eq(['true'])
+      end
+
+      it 'is an instance of matching parameterized URI' do
+        code = <<-CODE
+            $o = URI('https://example.com/a/b')
+            notice($o =~ URI[scheme => https, host => 'example.com'])
+        CODE
+        expect(eval_and_collect_notices(code)).to eq(['true'])
+      end
+
+      it 'is an instance of matching default URI' do
+        code = <<-CODE
+            $o = URI('https://example.com/a/b')
+            notice($o =~ URI)
+        CODE
+        expect(eval_and_collect_notices(code)).to eq(['true'])
+      end
+
+      it 'path is not matched by opaque' do
+        code = <<-CODE
+            $o = URI('urn:a:b:c')
+            notice($o =~ URI[path => 'a:b:c'])
+        CODE
+        expect(eval_and_collect_notices(code)).to eq(['false'])
+      end
+
+      it 'opaque is path not matched by path' do
+        code = <<-CODE
+            $o = URI('https://example.com/a/b')
+            notice($o =~ URI[opaque => '/a/b'])
+        CODE
+        expect(eval_and_collect_notices(code)).to eq(['false'])
+      end
+
+      it 'is not an instance unless parameters matches' do
+        code = <<-CODE
+            $o = URI('https://example.com/a/b')
+            notice($o =~ URI[scheme => http])
+        CODE
+        expect(eval_and_collect_notices(code)).to eq(['false'])
+      end
+
+      it 'individual parts of URI can be accessed using accessor methods' do
+        code = <<-CODE
+            $o = URI('https://bob:pw@example.com:8080/a/b?a=b#frag')
+            notice($o.scheme)
+            notice($o.userinfo)
+            notice($o.host)
+            notice($o.port)
+            notice($o.path)
+            notice($o.query)
+            notice($o.fragment)
+        CODE
```

Comment:
Is it also possible to access attributes as `$o['scheme']`? If so, there should be a test. If not - then that should be added.

Reply:
It is not possible and adding it is problematic. An instance of `Object` will always expose its attributes using functions and not through `[]` access. Having both is ambiguous because starting with [PUP-8012](https://tickets.puppetlabs.com/browse/PUP-8012) an `Object` also can implement the `#[]` function to enable other type of access (such as to the contained hash of node specific errors in an `ExecutionResult`).